### PR TITLE
Added "Draw fractal dispatching work to a thread pool" example

### PR DESCRIPTION
**Pull request by budziq**
fixes: https://github.com/brson/rust-cookbook/issues/102

Currently the example is marked as "no_run" as it leaves `output.png` in rust-cookbook cwd due to https://github.com/brson/rust-skeptic/issues/23 (there is a PR for it https://github.com/brson/rust-skeptic/pull/26)
Also on weaker machines the example in debug build might take 1-2s and I would not like to make the test even slower.

Also maybe we could add additional thumbnail for the output.png?
output

### CODE REVIEW

File: src/concurrency.md [hunk 1]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
 to terminate before returning from the closure that passed into [`crossbeam::scope`] function, meaning that
 you can reference data from the calling function.
 
+[ex-threadpool-fractal]: #ex-threadpool-fractal
+<a name="ex-threadpool-fractal"></a>
+## Draw fractal dispatching work to a thread pool
+
+[![threadpool-badge]][threadpool] [![num-badge]][num] [![num_cpus-badge]][num_cpus] [![image-badge]][image] [![cat-concurrency-badge]][cat-concurrency][![cat-science-badge]][cat-science][![cat-rendering-badge]][cat-rendering]
+
+Draws a fractal from [Julia set] to an image utilizing a thread pool for computation.
```
Comment (dtolnay) [inline src/concurrency.md]: Add a thumbnail here with a link to the full sized image.

File: src/concurrency.md [hunk 2]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+#
+# error_chain! {
+#     foreign_links {
+#         MpscRecv(RecvError);
+#     }
+# }
+
+// Function converting intensity values to RGB
+// Based on http://www.efg2.com/Lab/ScienceAndEngineering/Spectra.htm
+fn wavelength_to_rgb(wavelength: u32) -> Rgb<u8> {
```
Comment (dtolnay) [inline src/concurrency.md]: Hide this like we hide all the error handling code. This is not important to understanding threadpool.

File: src/concurrency.md [hunk 3]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+            break;
+        }
+        z = z * z + c;
+        i = t;
+    }
+    i
+}
+
+// Normalizes color intensity values within RGB range
+fn normalize(color: f32, factor: f32) -> u8 {
```
Comment (dtolnay) [inline src/concurrency.md]: Hide this too.

File: src/concurrency.md [hunk 4]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+        701...780 => 0.3 + 0.7 * (780. - wave) / (780. - 700.),
+        _ => 1.0,
+    };
+
+    let (r, g, b) = (normalize(r, factor), normalize(g, factor), normalize(b, factor));
+    Rgb::from_channels(r, g, b, 0)
+}
+
+// Maps Julia set distance estimation to intensity values
+fn julia(c: Complex<f32>, x: u32, y: u32, width: u32, height: u32, max_iter: u32) -> u32 {
```
Comment (dtolnay) [inline src/concurrency.md]: Hide this too.

File: src/concurrency.md [hunk 5]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+}
+
+fn run() -> Result<()> {
+    // resultant image size
+    let (width, height) = (800, 600);
+    let mut img = ImageBuffer::new(width, height);
+
+    let iterations = 300;
+    // precomputed pixel colors from possible intensity range
+    let color_lut: Vec<_> = (0..iterations)
```
Comment (dtolnay) [inline src/concurrency.md]: Remove this and inline the computation below. This seems like a performance optimization that has nothing to do with threadpool so it doesn't provide value to the example.

File: src/concurrency.md [hunk 6]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+                         tx.send((x, y, i)).expect("Could not send data!");
+                     });
+    }
+
+    for _ in 0..(width * height) {
+        let (x, y, i) = rx.recv()?;
+        let pixel = color_lut[i as usize];
+        img.put_pixel(x, y, pixel);
+    }
+    let _ = img.save("output.png");
```
Comment (dtolnay) [inline src/concurrency.md]: Show how to handle this error.

File: src/concurrency.md [hunk 7]
```diff
@@ -63,24 +64,170 @@ fn find_max(arr: &[i32], start: usize, end: usize) -> i32 {
...
+    i
+}
+
+// Normalizes color intensity values within RGB range
+fn normalize(color: f32, factor: f32) -> u8 {
+    ((color * factor).powf(0.8) * 255.) as u8
+}
+
+fn run() -> Result<()> {
+    // resultant image size
```
Comment (dtolnay) [inline src/concurrency.md]: Remove this comment. I think the names `width` and `height` and the next line are self-explanatory.

**Comment by budziq** [comment]
Rebased and fixed :)

Don't know how I've missed that Result on `ImageBuffer::save`. Maby we should have `![deny(warnings)]` somewhere in skeptic config (in such way we could be also asked to fix snippets on depreciations)

Also I did not know a better way to handle the thumbnail except for inline html pointing to github static content.

**Pull request merged**